### PR TITLE
Add authentication for Safekeeper

### PR DESCRIPTION
**Pull request by yeputons**
Based on #2730

### CODE REVIEW

File: control_plane/src/compute.rs [hunk 1]
```diff
@@ -306,6 +306,9 @@ impl PostgresNode {
         conf.append("shared_preload_libraries", "neon");
         conf.append_line("");
         conf.append("neon.pageserver_connstring", &pageserver_connstr);
+        if let AuthType::NeonJWT = auth_type {
+            conf.append("neon.safekeeper_token_env", "$ZENITH_AUTH_TOKEN");
```
Comment (yeputons) [inline control_plane/src/compute.rs:310]: This env variable is already set when running the compute node, it's used inside `pageserver_connstring`

**Comment by petuhovskiy** [comment]
Would be good to update https://github.com/neondatabase/neon/blob/main/docs/authentication.md, to follow current state of things better.

IIUC, we currently have 3 storage services operating with Neon JWT:
- compute
- safekeeper
- pageserver

I'd like to have documentation on what tokens are placed in the config of these services, and whether they're optional or not.

In safekeeper and pageserver we have libpq and http interfaces, and we accept queries with tokens on these interfaces. Clarification on what tokens we should use for these interfaces will be also nice.

This is not strictly related to this PR, but you seem to have context for these things which is needed to write a good doc :)
This doc will also make reviewing this PR a little easier.

File: libs/utils/src/auth.rs [hunk 2]
```diff
@@ -23,6 +23,7 @@ const JWT_ALGORITHM: Algorithm = Algorithm::RS256;
 pub enum Scope {
     Tenant,
     PageServerApi,
+    AllSafekeeperData,
```
Comment (arssher) [inline libs/utils/src/auth.rs:26]: Why All?

File: libs/utils/src/http/endpoint.rs [hunk 3]
```diff
@@ -137,10 +137,14 @@ pub fn auth_middleware<B: hyper::body::HttpBody + Send + Sync + 'static>(
     })
 }
 
-pub fn check_permission(req: &Request<Body>, tenant_id: Option<TenantId>) -> Result<(), ApiError> {
+pub fn check_permission_with<E: Display>(
```
Comment (arssher) [inline libs/utils/src/http/endpoint.rs:140]: E can/should be std::error:Error

File: control_plane/src/compute.rs [hunk 4]
```diff
@@ -324,6 +324,9 @@ impl PostgresNode {
...
```
Comment (arssher) [inline control_plane/src/compute.rs:328]: This won't be needed if we make $ZENITH_AUTH_TOKEN the default GUC value -- and better reuse it as a token to both  pageserver and safekeeper (can be in separate pr).

File: control_plane/src/storage.rs [hunk 5]
```diff
@@ -239,6 +240,13 @@ impl PageServerNode {
         let mut filled_cmd = fill_rust_env_vars(cmd.args(&args).arg("--daemonize"));
         filled_cmd = fill_aws_secrets_vars(filled_cmd);
 
+        if self.env.pageserver.auth_type != AuthType::Trust {
```
Comment (arssher) [inline control_plane/src/storage.rs:243]: worth a comment 'generate a token to connect from pageserver to safekeeper'

File: safekeeper/src/wal_service.rs [hunk 6]
```diff
@@ -5,25 +5,32 @@
...
 use crate::handler::SafekeeperPostgresHandler;
 use crate::SafeKeeperConf;
 use utils::postgres_backend::{AuthType, PostgresBackend};
 
 /// Accept incoming TCP connections and spawn them into a background thread.
-pub fn thread_main(conf: SafeKeeperConf, listener: TcpListener) -> Result<()> {
+pub fn thread_main(
+    conf: SafeKeeperConf,
+    listener: TcpListener,
+    auth: Option<Arc<JwtAuth>>,
```
Comment (arssher) [inline safekeeper/src/wal_service.rs:21]: Let's move public key inside `SafekeeperConf`.

**Review by arssher — APPROVED** [review]

File: libs/utils/src/http/endpoint.rs (hunk 3 above)
Comment (yeputons) [reply libs/utils/src/http/endpoint.rs:140]: We use `anyhow::Error` instead. It does not implement `std::error::Error` for some reason. Changed to an explicit type.

File: safekeeper/src/wal_service.rs (hunk 6 above)
Comment (yeputons) [reply safekeeper/src/wal_service.rs:21]: Will do together with corresponding Pageserver refactoring in a separate PR.

**Comment by yeputons** [comment]
@arssher @petuhovskiy Would you mind checking if the updated `authentication.md` makes sense? It looks very similar to a file describing data flow in Neon, if we have one.

File: docs/authentication.md [hunk 7]
```diff
@@ -1,30 +1,148 @@
...
+but may be not.
+
+### Safekeeper
+#### Overview
+Safekeeper keeps track of multiple tenants, each having multiple timelines.
+
+#### Incoming connections
+Safekeeper accepts connections from Compute/Pageserver, each
+connection corresponds to a specific timeline and requires
+a corresponding JWT token.
```
Comment (arssher) [inline docs/authentication.md:130]: Let mention here config param which specifies public key to check tokens with.

File: docs/authentication.md [hunk 8]
```diff
@@ -1,30 +1,148 @@
...
+but a non-rogue client never accesses another timeline even by an accident:
+timeline IDs are random and hard to guess.
 
-JWT authentication now supports two scopes: tenant and pageserverapi. Tenant scope is intended for use in tenant related api calls, e.g. create_branch. Compute launched for particular tenant also uses this scope. Scope pageserver api is intended to be used by console to manage pageserver. For now we have only one management operation - create tenant.
+#### Incoming connections
+All incoming connections are from PostgreSQL clients.
+Their authentication is just plain PostgreSQL authentication and out of scope for this document.
 
-Examples for token generation in python:
+There is no administrative API except for those provided by PostgreSQL.
```
Comment (arssher) [inline docs/authentication.md:64]: Word 'for' is redundant here at least.

File: docs/authentication.md [hunk 9]
```diff
@@ -1,30 +1,148 @@
...
+For each timeline, it connects to the corresponding Safekeeper.
+Information about "corresponding Safekeeper" is published by Safekeepers
+in the Etcd, but they do not publish access tokens, otherwise what is
+the point of authentication.
+
+Pageserver keeps a connection to some set of Safekeepers, which
+may or may not correspond to active Computes.
+
+_TODO_: describe specific situations when a Pageserver _needs_
+access to a Safekeeper while the Compute is down.
```
Comment (arssher) [inline docs/authentication.md:94]: I think this is out of scope here, but for example when compute finished earlier than pageserver consumed all WAL stream and it needs to catch up.

File: docs/authentication.md [hunk 10]
```diff
@@ -1,30 +1,148 @@
...
+
+_TODO_: what about pageserver replicas? Is it a thing?
```
Comment (arssher) [inline docs/authentication.md:96]: They are not different from main pageserver WRT auth.

File: docs/authentication.md [hunk 11]
```diff
@@ -1,30 +1,148 @@
...
-tenant_token = jwt.encode({"scope": "tenant", "tenant_id": ps.initial_tenant}, auth_keys.priv, algorithm="RS256")
+#### Incoming connections
+Pageserver listens for connections from computes.
+Each compute should present a token valid for the timeline's tenant.
+
+Pageserver also has HTTP API: some parts are per-tenant,
+some parts are server-wide, these are different scopes.
+
+The `auth_type` configuration variable in Pageserver's config may have
+either of three values:
```
Comment (petuhovskiy) [inline docs/authentication.md:106]: Typo?

```suggestion
either of two values:
```

**Comment by petuhovskiy** [comment]
Great doc, thanks @yeputons! Will review the code after the rebase, but it looks mostly ok.

File: pageserver/src/bin/pageserver.rs [hunk 12]
```diff
@@ -256,6 +258,23 @@ fn start_pageserver(conf: &'static PageServerConf) -> anyhow::Result<()> {
     };
     info!("Using auth: {:#?}", conf.auth_type);
 
+    match var("ZENITH_AUTH_TOKEN") {
```
Comment (petuhovskiy) [inline pageserver/src/bin/pageserver.rs:261]: Random thought, can we rename `ZENITH_AUTH_TOKEN` to `NEON_AUTH_TOKEN`? We don't use it extensively yet, should be easy to rename in all places.

File: libs/utils/src/auth.rs [hunk 13]
```diff
@@ -21,8 +21,15 @@ const JWT_ALGORITHM: Algorithm = Algorithm::RS256;
...
 #[serde(rename_all = "lowercase")]
 pub enum Scope {
+    // Provides access to all data for a specific tenant (specified in `struct Claims` below)
+    // TODO: join these two?
     Tenant,
+    // Provides blanket access to all tenants on the pageserver plus pageserver-wide APIs.
+    // Should only be used e.g. for status check/tenant creation/list.
     PageServerApi,
+    // Provides blanket access to all data on the safekeeper plus safekeeper-wide APIs.
+    // Should only be used e.g. for status check.
```
[the remark this answers is not in the document]
Comment (yeputons) [reply libs/utils/src/auth.rs:31]: Yes, at the moment, adding a comment.

**Comment by yeputons** [comment]
Follow-up PRs and TODOs:

* [ ] `ZENITH_AUTH_TOKEN` --> `NEON_AUTH_TOKEN`
* [ ] Move move public key inside `SafekeeperConf`/`PageserverConf` and load it on start instead of passing a separate argument through chains of calls (`auth_validation_public_key_path` --> `auth_validation_public_key`).
* [ ] `Claims`: make it more structured. E.g. make `Scope::Tenant, tenant_id=None` impossible.
* [ ] Think about generating tokens for debug purposes when SSHed to a staging machine. E.g. `curl http://localhost:12345/some-pageserver-api` is useful

File: pageserver/src/auth.rs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+use utils::auth::{Claims, Scope};
+use utils::id::TenantId;
+
+pub fn check_permission(claims: &Claims, tenant_id: Option<TenantId>) -> Result<()> {
+    match (&claims.scope, tenant_id) {
+        (Scope::Tenant, None) => {
+            bail!("Attempt to access management api with tenant scope. Permission denied")
+        }
+        (Scope::Tenant, Some(tenant_id)) => {
+            if claims.tenant_id.unwrap() != tenant_id {
```
[the remark this answers is not in the document]
Comment (yeputons) [reply pageserver/src/auth.rs:11]: I'd prefer a separate PR: merge `Claims`+`Scope` into a single `enum` so it's impossible to construct incorrect data. Does that work?

File: control_plane/src/compute.rs (hunk 4 above)
Comment (yeputons) [reply control_plane/src/compute.rs:326]: Finally decided to keep `NULL` the default as the current default is no-auth.

File: pageserver/src/auth.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply pageserver/src/auth.rs:11]: Should work, thanks!

**Comment by yeputons** [comment]
Waiting for #2863 to be reviewed, approved, and merged before this.

**Comment by yeputons** [comment]
Rebased on top of #2863 (which is to be merged as soon as tests pass), no big conflicts with either `main` or the other PR.

Will merge this right PR after #2863, if there are no objections.

**Pull request merged**